lib/pos_fee_estimator.go

```diff
@@ -155,5 +180,149 @@ func (posFeeEstimator *PoSFeeEstimator) mempoolFeeEstimate(
+	if err != nil {
+		return 0, errors.Wrap(err, "mempoolFeeEstimate: Problem computing txn fee")
+	}
+	return txnFee, nil
+}
+
+func computeFeeGivenTxnAndFeeRate(txn *MsgDeSoTxn, feeRateNanosPerKB uint64) (uint64, error) {
+	// Create a clone of the txn, so we don't modify the original.
+	txnClone, err := txn.Copy()
+	if err != nil {
+		return 0, errors.Wrap(err, "computeFeeGivenTxnAndFeeRate: Problem copying txn")
+	}
+	// Set the nonce to the maximum value if it's not already set.
+	if txnClone.TxnNonce == nil {
+		txnClone.TxnNonce = &DeSoNonce{
+			ExpirationBlockHeight: math.MaxUint64,
+			PartialID:             math.MaxUint64,
+		}
+	}
+	if txnClone.TxnNonce.ExpirationBlockHeight == 0 {
+		txnClone.TxnNonce.ExpirationBlockHeight = math.MaxUint64
+	}
+	if txnClone.TxnNonce.PartialID == 0 {
+		txnClone.TxnNonce.PartialID = math.MaxUint64
+	}
+
+	// Set the TxnFeeNanos to the maximum value.
+	txnClone.TxnFeeNanos = math.MaxUint64
+	txnFeeNanos, err := computeFeeRecursive(txnClone, feeRateNanosPerKB)
+	if err != nil {
+		return 0, errors.Wrap(err, "computeFeeGivenTxnAndFeeRate: Problem computing fee rate recursively")
+	}
+	return txnFeeNanos, nil
+}
+
+func computeFeeRecursive(txn *MsgDeSoTxn, feeRateNanosPerKB uint64) (uint64, error) {
+	// Get the length of bytes in the txn.
+	txnBytesNoSignature, err := txn.ToBytes(true)
+	if err != nil {
+		return 0, errors.Wrap(err, "computeFeeRecursive: Problem serializing txn")
+	}
+	const MaxDERSigLen = 74
+	txnBytesLen := uint64(len(txnBytesNoSignature)) + MaxDERSigLen
+
+	txnFee := txnBytesLen * feeRateNanosPerKB / 1000
+
+	if (txnBytesLen*feeRateNanosPerKB)%1000 != 0 {
+		txnFee++
+	}
+	if txnFee < txn.TxnFeeNanos {
+		txn.TxnFeeNanos = txnFee
+		return computeFeeRecursive(txn, feeRateNanosPerKB)
+	}
+	return txnFee, nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) estimateTxnFeeGivenTransactionRegister(
+	txn *MsgDeSoTxn,
+	txnRegister *TransactionRegister,
+	congestionFactorBasisPoints uint64,
+	priorityPercentileBasisPoints uint64,
+	maxBlockSize uint64,
+) (uint64, error) {
+	feeRateNanosPerKB, err := posFeeEstimator.estimateFeeRateNanosPerKBGivenTransactionRegister(txnRegister, congestionFactorBasisPoints, priorityPercentileBasisPoints, maxBlockSize)
+	if err != nil {
+		return 0, errors.Wrap(err, "estimateTxnFeeGivenTransactionRegister: Problem computing fee rate")
+	}
+	txnFee, err := computeFeeGivenTxnAndFeeRate(txn, feeRateNanosPerKB)
+	if err != nil {
+		return 0, errors.Wrap(err, "estimateTxnFeeGivenTransactionRegister: Problem computing txn fee")
+	}
+	return txnFee, nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) estimateFeeRateNanosPerKBGivenTransactionRegister(
+	txnRegister *TransactionRegister,
+	congestionFactorBasisPoints uint64,
+	priorityPercentileBasisPoints uint64,
+	maxBlockSize uint64,
+) (uint64, error) {
+	it := txnRegister.GetFeeTimeIterator()
```

Comment:
You need to grab the read lock on the transaction register before iterating over it. This makes sure that the transaction register doesn't shift from underneath you during iteration

```suggestion
	txnRegister.RLock()
	defer txnRegister.RUnlock()
	
	it := txnRegister.GetFeeTimeIterator()
```